### ISSUE TITLE
Password Protected Notes, Notebooks, and/or Content

### ISSUE THREAD

**Issue by jasondavis** [paperwork/paperwork#188]
A feature request I have is to be able to password protect either all of these or at least some of these:    
- Notebooks (all notes in a notebook, perhaps even the note names in that notebook)  
- Notes (password protect viewing the content of a note on a per note basis)  
- Content (password protect part of a note.  Evernote does this and the screenshot image below shows an example inside of evernote where a password is password protected from view)  

I currently use Evernote to store Bank account info, online accounts, bill accounts, all sorts of passwords and account numbers.  For this type of data, even running on a Paperwork install running on your own computer, it is still nice to be able to prevent prowling eyes who might have access to your computer from viewing some note content.

I realize that most likely this would only be hidden on the Frontend but that is really all I need.  For me it's more about people around me viewing my private data like passwords.

So would love to see something like this implemented into Paperwork!  Any thoughts, comments?

![2015-02-02_20-27-11](https://cloud.githubusercontent.com/assets/186818/6012773/3447b8ca-ab1a-11e4-9497-35dff3d7b06a.png)

Right click menu to reveal content, also double clicking will do the same  
![2015-02-02_20-37-17](https://cloud.githubusercontent.com/assets/186818/6012863/568d006a-ab1b-11e4-8dfd-4232cd861e85.png)

**Comment by devilx** [comment]
@jasondavis thank you for the detailed suggestion! Now, the first question here that pops into my mind would be: Should the note be hidden, until the user enters a password (super easy, super simple solution) or should the note be encrypted within the database using that password and not only be visually hidden, but actually unable to read without the correct password (hard but awesome solution).

We need to remember, that there a certain traps here. First of all the versioning. Every version would be be needed to be encrypted with the very same password and automatically decrypted when the note's password was provided correctly and the Wayback machine was launched.

Second, content_preview. What should be in there for an encrypted note?

Third, how should an encrypted note be created? Should we build a "New encrypted Note" in addition to "New Note", within the menu? Or should a subsequent encrypting of a standard note be possible? What would happen to all the versions that might have existed so far? Should these also be encrypted afterwards?

Fourth, attachments. Should attachments of a note also be saved encrypted on the filesystem? What about attachments that were uploaded before the note was encrypted, if subsequent encryption should be possible?

Fifth, attachment content. Should the attachment be OCR parsed if the note it was uploaded to was encrypted?

I could continue for a few hours like this. I guess, this is a *very* tricky topic and needs to be thought through very well upon implementation.

**Comment by jasondavis** [comment]
It sounds like a nightmare when considering all those issues!  Would like to hear some others feedback on it as well.    Looking at something like WordPress again as an example of how they handle per-post passwords.  They do not alter the post content but simple password protect the view of it on the frontend.  For my own personal needs and for simplicity sake, this would be enough for my own needs.

I do also like the method you mention of encrypting the note content but the trouble of then encrypting all versions of the note, previous and future just sounds like a mess and likely more issues that aren't being though of right now!  Sometimes the simple route might be the best route and this might be the case here.

As mentioned i'm interested to hear other feedback as well, thanks

**Comment by jamesrascal** [comment]
Personally I love this idea. @devilx  brings up a lot of good points. Something that I think would work for the moment is having the note content hidden until you type a password. 

I would like to have the option to possibly install Paperwork in a paranoid mode where you can set up a master key to encrypt all of the notes by default and then provide the user an option to setup their account as a normal or encrypted account. If set as an encrypted account they would need to use a secondary encryption password to open the account. This would keep any notes and revisions encrypted until it is decrypted by the user. On the admin panel there should be the option to enable or disable this feature. It would also need to be configured at the creation of the account.

**Comment by jasondavis** [comment]
@jamesrascal Yu bring up another possible use case however even if it was implemented in the manner you mention then I believe there should still be an option to password protect on a per note basis for us users who want to protect data behind a password but only on certain notes and not others.

**Comment by devilx** [comment]
@jasondavis let me clarify this:

WordPress does not encrypt the content inside the DB using the password but just not show you the content until you enter the right password. That's the easy way I described above, which could be implemented pretty quickly into Paperwork but would only provide a small level of security.

On the other hand, the "paranoid mode" (as called by @jamesrascal) would be the second way I described above – regardless of whether there's a master-key + per-user keys or just per-user keys. The master-key would probably only be good for allowing administrators to decrypt user notes in order to comply with government regulatory. Therefor, I'm not sure, whether we should provide this feature. ;-)

However, in addition to the issues mentioned above (with the "paranoid mode"), another one just came to my mind:

Sharing a notebook would be impossible, as the user the notebook (+ all notes inside) was shared to would need to key to decrypt the content (-> sharing user's password). The only way I could see here would be, to decrypt all the notes (+ all versions!) again and re-encrypt them using both user's passwords.

Honestly, I think the whole encryption thing is not worth the work, as people, who would really want to have their data fully encrypted on the back-end, would simply host their *SQL-server within an fully encrypted environment.

Then for securing the transmission, #22 would get implemented and we would be all good.

So basically this means: The notes table will need an additional "password" column, which can be NULL, where we store a password hash. The note's content isn't affected by this in any form. As soon, as the *password* column wouldn't be *NULL*, the API controllers would require an additional password to be passed inside the GET/POST requests. Only, if the password is the correct one, a GET would return the content and a POST would update it. As long as the password passed to the API doesn't match the saved one, some error would be returned.

**Comment by jasondavis** [comment]
@devilx The password column is really the way I prefer it to be and that's exactly how WOrdPress does it too.  A simple password to access the view note screen is all it needs to do and for me personally is plenty good enough!

**Comment by devilx** [comment]
Let's implement it the simple way. I've added this feature to Paperwork 2.0.

**Comment by mrusme** [comment]
I rethought the implementation of password protection recently and I'd object the decision once again by not implementing the *simple way* but rather have real cryptography in place. The main reason for this is that the simple way would lure users into believing their notes would be safe from hackers/government access/etc. while they in fact would not be.

Hence, please refer [to this issue](https://github.com/paperwork/paperwork/issues/891) for further updates on this topic.
